src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+            struct linger sl;
+            sl.l_onoff = (onoff == TRUE_ATOM) ? 1 : 0;
+            sl.l_linger = term_to_int(linger);
+            int res = setsockopt(fd_obj->fd, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl));
+            if (UNLIKELY(res != 0)) {
+
+                if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+                    AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                    RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+                }
+                term error_tuple = term_alloc_tuple(2, &ctx->heap);
+                term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+                term_put_tuple_element(error_tuple, 1, posix_errno_to_term(errno, global));
+                return error_tuple;
+            } else {
+                return OK_ATOM;
+            }
+            // TODO add more as needed
+            // int flag = 1;
+            // int res = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int));
+            // if (UNLIKELY(res != 0)) {
+            //     AVM_LOGW(TAG, "Failed to set TCP_NODELAY.");
+            // }
+        } else {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+//
+// sockname
+//
+
+static term nif_socket_sockname(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_sockname\n");
+    UNUSED(argc);
+
+    VALIDATE_VALUE(argv[0], term_is_socket);
+
+    GlobalContext *global = ctx->global;
+
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), get_socket(argv[0]), socket_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct SocketFd *fd_obj = (struct SocketFd *) fd_obj_ptr;
+    if (fd_obj->fd) {
+        struct sockaddr_in addr;
+        socklen_t addrlen = sizeof(addr);
+        int res = getsockname(fd_obj->fd, (struct sockaddr *) &addr, &addrlen);
+
+        if (UNLIKELY(res != 0)) {
+            AVM_LOGE(TAG, "Unable to getsockname: fd=%i res=%i.", fd_obj->fd, res);
+            if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+                AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+            }
+
+            term error_tuple = term_alloc_tuple(2, &ctx->heap);
+            term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+            term_put_tuple_element(error_tuple, 1, posix_errno_to_term(errno, global));
+            return error_tuple;
+        } else {
+            // {ok, #{addr => {a,b,c,d}, port => integer()}}
+            if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2) + term_map_size_in_terms(2) + TUPLE_SIZE(4)) != MEMORY_GC_OK)) {
+                AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+            } else {
```

Comment:
else is unnecessary here

Reply:
I realize that, but for some reason I prefer not to over-sue premature return.